### PR TITLE
[Reviewer: Matt] Sto1226

### PR DESCRIPTION
**Pull request by eleanor-merry**
Matt, can you please review these changes to hook the rt_ignore_dh extension when sending/receiving a message rather than when forwarding to a peer? I've also improved the check for whether the Destination-Realm is valid (previously if the Destination-Host was OK the request was accepted)

Tested with seagull

### CODE REVIEW

File: extensions/rt_ignore_dh/rt_ignore_dh.c [hunk 1]
```diff
@@ -165,8 +165,13 @@ static int stow_destination_host(struct msg **msg) {
...
 			CHECK_FCT(fd_msg_avp_add(pi_avp, MSG_BRW_LAST_CHILD, ps_avp));
 
-			/* remove Destination-Host from message */
-			fd_msg_free((msg_or_avp*)avp);
+			/* set the Destination-Host to the local Diameter Identity */
+			union avp_value dh;
+			memset(&dh, 0, sizeof(dh));
+			dh.os.data = memdup(fd_g_config->cnf_diamid, fd_g_config->cnf_diamid_len);
+			dh.os.len = fd_g_config->cnf_diamid_len;
+			CHECK_FCT(fd_msg_avp_setvalue(avp, &dh));
```
Comment (mirw) [inline extensions/rt_ignore_dh/rt_ignore_dh.c]: Doesn't this mean that the response will be sent from the local Diameter identity?  Surely that's wrong?

File: extensions/rt_ignore_dh/rt_ignore_dh.c [hunk 2]
```diff
@@ -177,27 +182,37 @@ static int stow_destination_host(struct msg **msg) {
 }
 
 /* The callback for putting Destination-Host into Proxy-Info and restoring it on the way back */
-static int rt_ignore_destination_host(void * cbdata, struct msg **msg) {
+static void rt_ignore_destination_host(enum fd_hook_type type,
```
Comment (mirw) [inline extensions/rt_ignore_dh/rt_ignore_dh.c]: I'm not sure we should change the existing module.  Shouldn't we create a new one?

File: extensions/rt_ignore_dh/rt_ignore_dh.c [hunk 3]
```diff
@@ -165,8 +165,13 @@ static int stow_destination_host(struct msg **msg) {
 			CHECK_FCT(fd_msg_avp_add(pi_avp, MSG_BRW_LAST_CHILD, ph_avp));
 			CHECK_FCT(fd_msg_avp_add(pi_avp, MSG_BRW_LAST_CHILD, ps_avp));
 
-			/* remove Destination-Host from message */
-			fd_msg_free((msg_or_avp*)avp);
+			/* set the Destination-Host to the local Diameter Identity */
```
Comment (mirw) [inline extensions/rt_ignore_dh/rt_ignore_dh.c]: Why do we need to replace Destination-Host here?  Why can't we just strip it?

File: libfdcore/routing_dispatch.c [hunk 4]
```diff
@@ -715,8 +715,15 @@ static int msg_rt_in(struct msg * msg)
 			return 0;
 		}
 
+		/* If the Destination-Realm is wrong */
```
Comment (mirw) [inline libfdcore/routing_dispatch.c]: Why this change?  We seem to have changed freeDiameter behavior!

**Comment by mirw** [comment]
Looks good.  A few comments.

File: extensions/rt_ignore_dh/rt_ignore_dh.c (hunk 3 above)
Comment (eleanor-merry) [inline extensions/rt_ignore_dh/rt_ignore_dh.c]: This causes problems later on where the request is rejected because it doesn't have a Destination-Host

Comment (mirw) [inline extensions/rt_ignore_dh/rt_ignore_dh.c]: OK.  I think that's a bug, but I'm happy that we can apply this change for now and raise an issue to investigate the absence of Destination-Host later.

File: extensions/rt_ignore_dh/rt_ignore_dh.c (hunk 2 above)
Comment (eleanor-merry) [inline extensions/rt_ignore_dh/rt_ignore_dh.c]: Ok

**Comment by eleanor-merry** [comment]
Matt, I've updated this PR to use a new PRE_SEND hook - can you please do a re-review of this?

**Comment by mirw** [comment]
Looks good.  (As discussed, I'm not keen on the ROUTING_ERROR string-matching in cpp-common, but we're dealing with that separately.)
